### ISSUE TITLE
Re-consider `MERA` and `TEBD`

### ISSUE THREAD

**Issue by fliingelephant** [nzy1997/ParametricDFT.jl#66]
Phase gates (diagonal entries only, each has magnitude 1) without non-diagonal gates followed actually DO NOT contribute any to the loss.

Hence, `TEBD` and `MERA` are merely several Hadamard's. After training, they become single-qubit gates, and the phase gates do not evolve at all, or just accumulate numerical errors.

**Comment by zazabap** [comment]
Thanks for catching this — the analysis is correct.

Both TEBD and MERA have structure `U = U_phase ∘ U_hadamard`, where all controlled-phase gates are diagonal with unit-magnitude entries (`diag(1, 1, 1, e^{iθ})`). Since our losses (L1 and MSE) depend only on magnitudes:

- **L1**: `∑|U·x| = ∑|U_H·x|` — phase gates preserve magnitudes, so `|e^{iφ(θ)}| = 1` kills any dependence on θ. `∂L/∂θ = 0` exactly.
- **MSE**: `topk_truncate` selects by magnitude (θ-independent), and phase gates cancel in the round-trip: `U_H⁻¹ · U_phase⁻¹ · U_phase · topk(U_H · x) = U_H⁻¹ · topk(U_H · x)`.

So the phase parameters have zero effective gradient — TEBD/MERA reduce to fixed Hadamard transforms with no learnable parameters.

The root cause is that all phase gates sit **after** the last non-diagonal operation (Hadamard), with nothing downstream to convert phase differences into magnitude differences. Compare with EntangledQFT's `:middle` mode, which interleaves entanglement gates between Hadamard gates — there the phases do affect magnitudes because subsequent H gates mix them in.

**Fix direction**: Add non-diagonal operations (single-qubit rotations like Ry) after the phase gate layers, so that phase differences get converted into magnitude differences that the loss can see. This preserves the TEBD ring / MERA hierarchical connectivity while making the parameters actually trainable.

Will open a branch with the fix.

**Comment by fliingelephant** [comment]
Tried adding Hadamard's after phase gates (just like sandwiching). Tested on Quick Draw `moderate` with `shuffle=false` and `lr=0.01`.
| Circuit + Loss | Gates | 5% PSNR | 10% PSNR | 15% PSNR | 20% PSNR | 20% SSIM |
  |---|---|---|---|---|---|---|
  | FFT | — | 15.27 | 17.27 | 18.99 | 20.50 | 0.526 |
  | DCT | — | 16.08 | 18.43 | 20.39 | 22.13 | 0.543 |
  | QFT + Entropy | 30 | 16.50 | 19.24 | 21.65 | 23.89 | 0.756 |
  | EntangledQFT + Entropy | 35 | 16.50 | 19.24 | 21.66 | 23.89 | 0.756 |
  | QFT + MSE | 30 | 17.95 | 22.22 | 26.01 | 29.89 | 0.953 |
  | EntangledQFT + MSE | 35 | 17.95 | 22.22 | 26.00 | 29.90 | 0.954 |
  | **TEBD + Entropy** | **30** | **18.37** | 23.09 | 27.54 | 32.09 | 0.962 |
  | **TEBD + MSE** | **30** | 17.43 | **23.43** | **29.33** | **35.49** | **0.987** |

**Comment by zazabap** [comment]
@fliingelephant fabulous results! Is there explicit image drawing out like issue #65 ? Especially for TEBD+MSE, I am quite interested in how it works.

**Comment by fliingelephant** [comment]
@zazabap 
<img width="1904" height="2540" alt="Image" src="https://github.com/user-attachments/assets/3e2f84b1-4cee-43a9-8a0b-9d471e7c6ff0" />

<img width="1904" height="2540" alt="Image" src="https://github.com/user-attachments/assets/84db73cc-91dd-403b-aa92-1eaf598a332d" />

**Comment by fliingelephant** [comment]
For DIV2K 8q several methods converge to nearly identical result, and I think it's because we keep top 10% with MSE error. Now rerunning with top 20% and `heavy`. Maybe we better use entropy loss and learning rate 1e-2?

**Comment by fliingelephant** [comment]
DIV2K 8q

MSE + top10%, `moderate`
  | Circuit | 5% PSNR | 10% PSNR | 15% PSNR | 20% PSNR | 20% SSIM |
  |---|---|---|---|---|---|
  | FFT | 21.71 | 23.19 | 24.29 | 25.26 | 0.744 |
  | DCT | 22.33 | 24.00 | 25.32 | 26.51 | 0.781 |
  | QFT | 22.44 | 24.52 | 26.21 | 27.76 | 0.845 |
  | Entangled QFT | 22.44 | 24.52 | 26.21 | 27.76 | 0.846 |
  | TEBD | 22.45 | 24.52 | 26.21 | 27.76 | 0.846 |
  | MERA | 22.44 | 24.51 | 26.20 | 27.75 | 0.844 |

<img width="1600" height="1000" alt="Image" src="https://github.com/user-attachments/assets/c6ad9c6e-0be1-41bb-b4bb-387eee98a3ea" />

<img width="1600" height="1000" alt="Image" src="https://github.com/user-attachments/assets/c46e4c68-1082-4979-ae12-1cb510c6906e" />

**Comment by fliingelephant** [comment]
DIV2K 8q

MSE + top20% , `heavy`
| Circuit | 5% PSNR | 10% PSNR | 15% PSNR | 20% PSNR | 20% SSIM |
  |---|---|---|---|---|---|
  | FFT | 21.87 | 23.29 | 24.37 | 25.32 | 0.740 |
  | DCT | 22.49 | 24.11 | 25.41 | 26.60 | 0.776 |
  | QFT | 22.74 | 24.93 | 26.73 | 28.40 | 0.857 |
  | Entangled QFT | 22.81 | 25.12 | 27.00 | 28.75 | 0.863 |
  | TEBD | 22.82 | 25.12 | 27.00 | 28.75 | 0.863 |
  | MERA | 22.74 | 24.93 | 26.73 | 28.40 | 0.857 |

<img width="1600" height="1000" alt="Image" src="https://github.com/user-attachments/assets/4461bed6-9d43-4e2c-a5a3-e635c94bdc5b" />

<img width="1600" height="1000" alt="Image" src="https://github.com/user-attachments/assets/37a7693f-bdd4-4536-be50-c44257bd74a9" />

**Comment by zazabap** [comment]
The training loss is a bit weird, There should be a version based on the steps. Also, are you using quickdraw dataset everytime? What about the performance on other dataset? Did you add Hadamard gate after the TEBD this time? Or simply use the implementation on main branch.

**Comment by fliingelephant** [comment]
> The training loss is a bit weird, There should be a version based on the steps. Also, are you using quickdraw dataset everytime? What about the performance on other dataset? Did you add Hadamard gate after the TEBD this time? Or simply use the implementation on main branch.

These two are for DIV2K 8q
Hadamard gates added after TEBD see `hh/batched` branch

**Comment by zazabap** [comment]
Thanks a lot, I see the branch and will check.
